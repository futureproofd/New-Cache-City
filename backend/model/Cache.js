const mongoose = require('mongoose');
const mongoosePaginate = require('mongoose-paginate-v2');
const validator = require('validator');

const { Schema } = mongoose;

mongoose.Promise = global.Promise;

const mongodbErrorHandler = require('mongoose-mongodb-errors');

const cacheSchema = new Schema({
  name: {
    type: String,
    unique: true,
    lowercase: true,
    trim: true,
    validate: [validator.isAscii, 'Invalid Cache name'],
    required: 'Please provide a name for this Cache',
  },
  description: {
    type: String,
    unique: false,
    validate: [validator.isAscii, 'Invalid characters'],
    required: 'Please provide a description of the Cache',
  },
  createDate: {
    type: Date,
    default: Date.now,
  },
  location: {
    type: {
      type: String,
      default: 'Point',
    },
    coordinates: [
      {
        type: Number,
        required: 'Coordinates required for Cache',
      },
    ],
    address: {
      type: String,
      required: 'You must supply a general location',
    },
  },
  photo: String,
  author: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: 'User required',
  },
});

cacheSchema.plugin(mongodbErrorHandler);
cacheSchema.plugin(mongoosePaginate);

module.exports = mongoose.model('Cache', cacheSchema);
