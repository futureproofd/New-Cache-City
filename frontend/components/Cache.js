/* eslint-disable consistent-return */
/* eslint-disable no-underscore-dangle */
import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import CacheStyle from '../styles/CacheStyle';

const Cache = ({ item, user }) => {
  const renderCaches = () => {
    const owner = user._id === item.author;

    switch (owner) {
      case false:
        return (
          <Fragment>
            <Link to={{ pathname: '/cache', state: { item } }}>
              <h1>{item.name}</h1>
            </Link>
            <p>{item.description}</p>
            <img src={item.photo} alt={item.name} />
          </Fragment>
        );
      default:
        if (user._id === item.author) {
          return (
            <Fragment>
              <Link to={{ pathname: '/cache', state: { item, user } }}>
                <h1>Edit</h1>
              </Link>
              <Link to={{ pathname: '/cache', state: { item } }}>
                <h1>{item.name}</h1>
              </Link>
              <p>{item.description}</p>
              <img src={item.photo} alt={item.name} />
            </Fragment>
          );
        }
    }
  };

  return <CacheStyle key={item.id}>{renderCaches()}</CacheStyle>;
};

export default Cache;
