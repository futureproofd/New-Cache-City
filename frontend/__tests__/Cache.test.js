/* eslint-disable no-underscore-dangle */
/* eslint-disable no-undef */
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import React from 'react';
import Cache from '../components/Cache';

const fakeCache = {
  _id: 'someObjId',
  name: 'fake cache',
  description: 'testing 123',
  image: 'cache.jpg',
  createDate: '20191128',
  location: {
    coordinates: [{ lat: '123' }, { lng: '456' }],
    address: '123 fake street',
  },
  photo: 'cake.jpg',
  author: 'someObjId',
};

describe('<Cache />', () => {
  it('Renders and matches the snapshot', () => {
    const wrapper = shallow(<Cache item={fakeCache} user={fakeCache._id} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});

describe('<Cache />', () => {
  it('Renders the name', () => {
    const wrapper = shallow(<Cache item={fakeCache} user={fakeCache._id} />);
    expect(wrapper.find('Link h1').text()).toBe(fakeCache.name);
  });
});
