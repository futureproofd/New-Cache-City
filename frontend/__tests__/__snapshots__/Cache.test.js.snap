// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cache /> Renders and matches the snapshot 1`] = `
<styled.div>
  <Link
    to={
      Object {
        "pathname": "/cache",
        "state": Object {
          "item": Object {
            "_id": "someObjId",
            "author": "someObjId",
            "createDate": "20191128",
            "description": "testing 123",
            "image": "cache.jpg",
            "location": Object {
              "address": "123 fake street",
              "coordinates": Array [
                Object {
                  "lat": "123",
                },
                Object {
                  "lng": "456",
                },
              ],
            },
            "name": "fake cache",
            "photo": "cake.jpg",
          },
        },
      }
    }
  >
    <h1>
      fake cache
    </h1>
  </Link>
  <p>
    testing 123
  </p>
  <img
    alt="fake cache"
    src="cake.jpg"
  />
</styled.div>
`;
